harness/harness-core
# fix: [PL-32111]: Query optimization for list projects call

## Describe your changes
Project list call depends on fetching the list of permitted projects. Earlier the criteria was formed in the following way:
```
{
  "accountIdentifier": "OzzP3vZfT2qXhoCa_fy84w",
  "$or": [
    {
      "orgIdentifier": "cypress_org_Dinesh_DeBruijn",
      "identifier": "cypress_project_Magdalena_Meyer"
    },
    {
      "orgIdentifier": "cypress_org_Dinesh_DeBruijn",
      "identifier": "main_cypress_project_Hong_Jimenez"
    },
    {
      "orgIdentifier": "cypress_org_Unnur_Dijkstra",
      "identifier": "cypress_project_Steinunn_Mitchell"
    }
  ]
}
```
With increase in number of projects, the `$or` clause after having more than 100 entries ignores index and hence the query was not performant resulting into timeouts as well. This solved by querying on _id field as below:
```
{
  "$match": {
    "_id": {
      "$in": [
        "ObjectId("641847b742fd232b50e3aeb1"),
        "ObjectId("6418558c42fd232b50e477a4"),
        "ObjectId("641959e422533020bc2c6976"),
        "ObjectId("6418a3a15744f53d153cd00b"),
        "ObjectId("6418482e42fd232b50e3b2d3")
      ]
    }
  }
}
```
The above query uses `_id_` index and is far better performant.

**Testing:**
1. Create projects in shared org.
2. Created shared projects in non shared org
3. Created projects not shared with anyone.
4. Logged in with user to which projects are share.
5. Logged in user should be able to view only those projects which are shared with the user.

## Checklist

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/47070)
